libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/MapboxTripNotification.kt

```diff
@@ -1,29 +1,413 @@
-    override fun getNotification(): Notification {
-        TODO("not implemented")
+class MapboxTripNotification(
+    private val applicationContext: Context,
+    private val tripNavigationOptions: TripNavigationOptions,
+    private var currentRoute: Route
+) : TripNotification {
+
+    private val END_NAVIGATION_ACTION = "com.mapbox.intent.action.END_NAVIGATION"
+    private val SET_BACKGROUND_COLOR = "setBackgroundColor"
+
+    private var notificationManager: NotificationManager? = null
+    private lateinit var notification: Notification
+    private var collapsedNotificationRemoteViews: RemoteViews? = null
+    private var expandedNotificationRemoteViews: RemoteViews? = null
+    private var currentDistanceText: SpannableString? = null
+    private var instructionText: String? = null
+    private var currentManeuverId: Int = 0
+    private var isTwentyFourHourFormat: Boolean = false
+    private var etaFormat: String = ""
+    private var pendingOpenIntent: PendingIntent? = null
+    private var pendingCloseIntent: PendingIntent? = null
+    private var distanceFormatter: DistanceFormatter? = null
+
+    private val endNavigationBtnReceiver = object : BroadcastReceiver() {
+        override fun onReceive(applicationContext: Context, intent: Intent) {
+            // this@MapboxNavigationNotification.onEndNavigationBtnClick()
+        }
     }
 
-    override fun getNotificationId(): Int {
-        TODO("not implemented")
+    companion object {
+        private const val NAVIGATION_NOTIFICATION_ID = 5678
+        private const val NAVIGATION_NOTIFICATION_CHANNEL = "NAVIGATION_NOTIFICATION_CHANNEL"
+    }
+
+    init {
+        etaFormat = applicationContext.getString(R.string.eta_format)
+        initializeDistanceFormatter()
+        notificationManager = applicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
+        isTwentyFourHourFormat = DateFormat.is24HourFormat(applicationContext)
+
+        pendingOpenIntent = createPendingOpenIntent()
+        pendingCloseIntent = createPendingCloseIntent()
+
+        registerReceiver()
+        createNotificationChannel()
+        if (!::notification.isInitialized) {
+            notification = buildNotification()
+        }
     }
 
+    override fun getNotification(): Notification = notification
+
+    override fun getNotificationId(): Int = NAVIGATION_NOTIFICATION_ID
+
     override fun updateRouteProgress(routeProgress: RouteProgress?) {
-        TODO("not implemented")
+        routeProgress?.let { progress ->
+            updateNotificationViews(progress)
+            rebuildNotification()
+        }
     }
 
     override fun updateLocation(rawLocation: Location, enhancedLocation: Location) {
         TODO("not implemented")
     }
 
     override fun onTripSessionStopped(context: Context) {
-        TODO("not implemented")
+        unregisterReceiver(context)
+    }
+
+    // Package private (no modifier) for testing purposes
```

Comment:
This usually means an issue in the class design. Could you review this code and verify if dependencies can be injected better, or maybe logic can be delegated into smaller classes with limited responsibilities that then could be tested without unnecessarily exposing APIs?